src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -628,44 +615,44 @@ internal MemberAccessor MemberAccessorStrategy
-                // .NET 6 backward compatibility; use fallback to reflection serialization
-                // TODO: Consider removing this behaviour (needs to be filed as a breaking change).
-                _effectiveJsonTypeInfoResolver = JsonTypeInfoResolver.Combine(ctx, DefaultJsonTypeInfoResolver.RootDefaultInstance());
-            }
-            else
+            if (!_isInitializedForReflectionSerializer)
             {
+                IsLockedInstance = true;
                 _typeInfoResolver ??= DefaultJsonTypeInfoResolver.RootDefaultInstance();
-            }
+                InitializeCachingContext();
 
-            if (_cachingContext != null && _cachingContext.Options != this)
-            {
-                // We're using a shared caching context deriving from a different options instance;
-                // for coherence ensure that it has been opted in for reflection-based serialization as well.
-                _cachingContext.Options.InitializeForReflectionSerializer();
-            }
+                if (_cachingContext.Options != this)
+                {
+                    // We're using a shared caching context deriving from a different options instance;
+                    // for coherence ensure that it has been opted in for reflection-based serialization as well.
+                    _cachingContext.Options.InitializeForReflectionSerializer();
+                }
 
-            IsInitializedForReflectionSerializer = true;
+                _isInitializedForReflectionSerializer = true;
+            }
         }
 
-        internal bool IsInitializedForMetadataGeneration { get; private set; }
+        private volatile bool _isInitializedForReflectionSerializer;
+
         internal void InitializeForMetadataGeneration()
         {
-            IJsonTypeInfoResolver? resolver = _effectiveJsonTypeInfoResolver ?? _typeInfoResolver;
-            if (resolver == null)
+            if (!_isInitializedForMetadataGeneration)
             {
-                ThrowHelper.ThrowInvalidOperationException_JsonTypeInfoUsedButTypeInfoResolverNotSet();
-            }
+                if (_typeInfoResolver is null)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_JsonTypeInfoUsedButTypeInfoResolverNotSet();
+                }
 
-            _isLockedInstance = true;
-            IsInitializedForMetadataGeneration = true;
+                IsLockedInstance = true;
+                _isInitializedForMetadataGeneration = true;
+            }
         }
 
-        private JsonTypeInfo? GetTypeInfoInternal(Type type)
+        private volatile bool _isInitializedForMetadataGeneration;
+
+        private JsonTypeInfo? GetTypeInfoNoCaching(Type type)
         {
-            IJsonTypeInfoResolver? resolver = _effectiveJsonTypeInfoResolver ?? _typeInfoResolver;
-            JsonTypeInfo? info = resolver?.GetTypeInfo(type, this);
+            JsonTypeInfo? info = _typeInfoResolver?.GetTypeInfo(type, this);
```

Comment:
consider adding Debug.Assert to make sure `_typeInfoResolver` is not null. I don't believe we should ever be here when it's null